### PR TITLE
Fix data pipeline and training loop bugs

### PR DESCRIPTION
**Pull request by tom-m-white**
Here is all of the stuff I changed:

*   **Added Missing FEN Functions:** Implemented `get_piece_at_square` so the data loader could figure out what piece was moving.
*   **Fixed Knight Move Encoding:** The move encoder was using the wrong logic for knights. I changed the `if` condition to correctly detect them.
*   **Synced Up Policy Size:** The `dataset.py` and `move_encoder.py` files disagreed on the total policy size (4608 vs 4672). I made the data loader import the correct size from the encoder so they're always in sync.
*   **Fixed Value Loss Shape:** Added `.unsqueeze(1)` to the value target so its shape would match the model's output, getting rid of the PyTorch warning.

The script should now train without crashing altho we need to implement the checkpoints.

### CODE REVIEW

File: src/python/hyperion_nn/config.py [hunk 1]
```diff
@@ -8,12 +8,14 @@
 import os
 import math
 
-
+SAVE_CHECKPOINT_EVERY_N_STEPS = 5000  # How often to save a model checkpoint
+VALIDATE_EVERY_N_STEPS = 1000         # How often to run validation
```
Comment (lj-dunphy) [inline src/python/hyperion_nn/config.py]: all constants should be in a class, these are vars in the TrainingConfig class. Either replace them with these there or remove these and use the ones in the TrainingConfig class

File: src/python/hyperion_nn/data_utils/dataset.py
Comment (lj-dunphy) [inline src/python/hyperion_nn/data_utils/dataset.py:1]: all looks good to me, good changes

File: src/python/hyperion_nn/data_utils/fen_parser.py [hunk 2]
```diff
@@ -123,16 +159,20 @@ def fen_to_nn_input(fen: str) -> np.ndarray:
     return nn_input
 
 
-
-def get_turn(fen_str: str) -> constants.Piece:
+def get_turn(fen_str: str) -> str:
```
Comment (lj-dunphy) [inline src/python/hyperion_nn/data_utils/fen_parser.py]: please try to use the constants.Piece enum class instead. it is much more readable and it leads to better error handling. if the class was broken, then try to fix it there instead of hard coding strings

File: src/python/hyperion_nn/data_utils/move_encoder.py
Comment (lj-dunphy) [inline src/python/hyperion_nn/data_utils/move_encoder.py:1]: same here, please try to use the constants.Piece enum class to make it more readable

File: src/python/hyperion_nn/training/train.py [hunk 3]
```diff
@@ -94,7 +94,7 @@ def train_model():
         loss_policy = policy_loss_fn(policy_logits, policy_loss_indices)
 
         value_prediction = torch.tanh(value_output)
-        loss_value = value_loss_fn(value_prediction, value_target)
+        loss_value = value_loss_fn(value_prediction, value_target.unsqueeze(1))
```
Comment (lj-dunphy) [inline src/python/hyperion_nn/training/train.py]: explain this more. the input should be a tensor. if it isnt working, why? this should ideally be something that is changed in the Dataset class and not in here

**Review by lj-dunphy — CHANGES_REQUESTED** [review]
See comments for changes. Mostly consistency stuff to remove hardcoded values and best practices to continue to go with the established program architecture/work flow

File: src/python/hyperion_nn/config.py (hunk 1 above)
Comment (tom-m-white) [reply src/python/hyperion_nn/config.py]: I went with the removing option, and we can just use the ones in the TrainingConfig class.

File: src/python/hyperion_nn/data_utils/fen_parser.py (hunk 2 above)
Comment (tom-m-white) [reply src/python/hyperion_nn/data_utils/fen_parser.py]: you cannot change the return type to constants.Piece because this enum represents chess pieces, like a pawn or a queen, not the player whose turn it is to move. to make it more readable instead, ill change the function's return type to a boolean (True for White, False for Black). this actually aligns with the constants used already, (constant.WHITE and constants.BLACK definitions).

File: src/python/hyperion_nn/training/train.py (hunk 3 above)
Comment (tom-m-white) [reply src/python/hyperion_nn/training/train.py]: ok there are 2 important parts here and it mainly boils down to: value_prediction vs value_target. 

value_prediction: it preserves the batch dimension which is a 2D tensor. specifically [batch_size, 1]. 

value_target: the ChessDataset's __getitem__  reads a game result: (like 0.0, -1.0, 1.0...) When the DataLoader takes batch_size of these single numbers, it stacks them together into a simple, 1D list of values. This makes the resulting shape a 1d tensor. specifically [batch_size].

PyTorch sees this and issues a UserWarning because while it can calculate the loss through a process called broadcasting, its often a sign that the user made a mistake.

the .unsqueeze(1) part makes it so pytorch adds another dimension at 1.

for example: 

     Before: it had a 1D tensor like tensor([1.0, -1.0, 0.0, ...]) with shape [64].
     After: it has a 2D tensor like tensor([[ 1.0], [-1.0], [ 0.0], ...]) with shape [64, 1]

now, both the prediction and the target have the exact same shape, which makes pytorch happy.

as you pointed out, the training loop shouldnt have to worry about cleaning up data shapes. The Dataset should provide data in the exact format the model and loss functions expect. so I'm going to change it so when the DataLoader collects a batch of these items, it will see that each value_target is already an array-like object of shape (1,). When it stacks batch_size of them together, it will create a final tensor of shape [batch_size, 1], which is exactly what we need here.

File: src/python/hyperion_nn/data_utils/move_encoder.py
Comment (tom-m-white) [reply src/python/hyperion_nn/data_utils/move_encoder.py:1]: made the change!! (see previous reply regarding not make it a constant.piece, and using constant.WHITE and constant.BLACK)

File: src/python/hyperion_nn/data_utils/fen_parser.py (hunk 2 above)
Comment (lj-dunphy) [reply src/python/hyperion_nn/data_utils/fen_parser.py]: sounds good, my bad for the error

File: src/python/hyperion_nn/training/train.py (hunk 3 above)
Comment (lj-dunphy) [reply src/python/hyperion_nn/training/train.py]: 🪐

File: src/python/hyperion_nn/training/train.py [hunk 4]
```diff
@@ -10,16 +10,35 @@
...
+root_logger.addHandler(file_handler)
+
+# 3. Create the console handler. This handler prints messages to the console
+console_handler = logging.StreamHandler(sys.stdout)
+# Set a higher level for the console. WARNING means it will only print messages
+# that are warnings, errors, or critical. INFO and DEBUG messages will be ignored.
+console_handler.setLevel(logging.WARNING) 
+console_formatter = logging.Formatter("[%(levelname)-8s] %(message)s")
+console_handler.setFormatter(console_formatter)
+root_logger.addHandler(console_handler)
```
Comment (lj-dunphy) [inline src/python/hyperion_nn/training/train.py:42]: just curious, why did you change this? is it just to look better, or is this more efficient? also, i think it is still important to have info and debug output to the console for debugging purposes, we should have at least one of those output to console

File: src/python/hyperion_nn/training/train.py [hunk 5]
```diff
@@ -72,64 +92,99 @@ def train_model():
...
+    ||_  $$_/   /$$__  $$|____  $$| $$| $$__  $$| $$| $$__  $$ /$$__  $$        |@ @ @ - @ @ - - - - @ @ @ = @ @ = @ @ @ @ @ @ @|
+    |  | $$    | $$  \__/ /$$$$$$$| $$| $$  \ $$| $$| $$  \ $$| $$  \ $$        |@ @ @ - - - - - @ @ @ @ @ - @ @ = @ @ @ @ @ @ @|
+    |  | $$ /$$| $$      /$$__  $$| $$| $$  | $$| $$| $$  | $$| $$  | $$        |@ @ @ @ - = - @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @|
+    |  |  $$$$/| $$     |  $$$$$$$| $$| $$  | $$| $$| $$  | $$|  $$$$$$$        |@ @ @ * * + + @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @|
+    |   \___/  |__/      \_______/|__/|__/  |__/|__/|__/  |__/ \____  $$        |@ @ * * * @ + + @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @|
+    |                                                          /$$  \ $$        |@ * * * @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @|
+    |                                                         |  $$$$$$/        |@ @ * @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @|
+    |                                                          \______/         |@ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @ @|
+    +-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-+
+    """)
```
Comment (lj-dunphy) [inline src/python/hyperion_nn/training/train.py]: can we keep this string in the constants file so it doesn't take up a ton of room in the file?

File: src/python/hyperion_nn/training/train.py [hunk 6]
```diff
@@ -72,64 +92,99 @@ def train_model():
...
+
+    for epoch in range(start_epoch, total_epochs):
```
Comment (lj-dunphy) [inline src/python/hyperion_nn/training/train.py:106]: consider making the progress bar steps instead of epochs, as the epochs will take a very long time

Comment (lj-dunphy) [reply src/python/hyperion_nn/training/train.py:106]: nvm just read the rest of the code

**Review by lj-dunphy — CHANGES_REQUESTED** [review]
small stuff

File: src/python/hyperion_nn/config.py [hunk 7]
```diff
@@ -53,7 +51,7 @@ class TrainingConfig:
 
     # ~~ Logging/Checkpointing ~~
     SAVE_CHECKPOINTS_EVERY_N_STEPS = 10_000  # save a checkpoint every N training steps
-    VALIDATE_EVERY_N_STEPS = 5 * SAVE_CHECKPOINTS_EVERY_N_STEPS
```
Comment (lj-dunphy) [inline src/python/hyperion_nn/config.py:53]: ?

Comment (tom-m-white) [reply src/python/hyperion_nn/config.py:53]: there were some errors with validate_every_n_steps and so I hardcoded a number just for testing. I will revert it back to:

VALIDATE_EVERY_N_STEPS = 5 * SAVE_CHECKPOINTS_EVERY_N_STEPS

File: src/python/hyperion_nn/training/train.py (hunk 4 above)
Comment (tom-m-white) [reply src/python/hyperion_nn/training/train.py:42]: yeahhhh, good point, ive changed it so it logs everything to both the console plus the log file. the steps and loss numbers now go into data/steps/ (this steps folder is created automatically if not already created). this way it keeps both the console and the logs clean!

File: src/python/hyperion_nn/training/train.py (hunk 5 above)
Comment (tom-m-white) [reply src/python/hyperion_nn/training/train.py]: solid suggestion, now it is print(constants.TRAINING_HEADER_ART). TRAINGING_HEADER_ART is stored in the constants.py file in utils

File: src/python/hyperion_nn/training/train.py (hunk 4 above)
Comment (tom-m-white) [reply src/python/hyperion_nn/training/train.py:42]: to answer your first part of the question. it allows me separate what was going to the log and what was going to the console. although with the recent changes that doesn't really matter as much now that we are storing things differently.

**Review by tom-m-white — COMMENTED** [review]
how does it look now?

**Comment by lj-dunphy** [comment]
LGTM! 🚀
